plugins/scale/scale-title-overlay.cpp

```diff
@@ -0,0 +1,309 @@
+    }
+};
+
+scale_show_title_t::scale_show_title_t() :
+    view_filter{[this] (auto)
+    {
+        update_title_overlay_opt();
+    }},
+
+    scale_end{[this] (wf::signal_data_t*)
+    {
+        show_view_title_overlay = title_overlay_t::NEVER;
+        last_title_overlay = nullptr;
+        mouse_update.disconnect();
+    }
+},
+
+add_title_overlay{[this] (wf::signal_data_t *data)
+    {
+        const std::string& opt = show_view_title_overlay_opt;
+        if (opt == "never")
+        {
+            /* TODO: support changing this option while scale is running! */
+            return;
+        }
+
+        auto signal = static_cast<scale_transformer_added_signal*>(data);
+        auto tr     = signal->transformer;
+        auto view   = tr->get_transformed_view();
+        auto ol     = std::make_shared<view_title_overlay_t>(*tr, title_font_size,
+            bg_color, text_color);
+        auto cb = std::make_shared<wf::scale_view_overlay>([ol, this, view] (
+            const wf::framebuffer_t& fb,
+            const wf::region_t& damage)
+        {
+            if (view->parent && (ol->pos != view_title_overlay_t::position::CENTER))
+            {
+                /*
+                 * Do not render dialogs' titles.
+                 * TODO: we still need a better solution when the overlay is
+                 * rendered in the center, since the dialog will be on top of it.
+                 * Also, handle the case when the dialog is larger than the
+                 * parent by extra padding.
+                 */
+                return;
+            }
+
+            auto tmp_view = view;
+            while (tmp_view->parent)
+            {
+                tmp_view = tmp_view->parent;
+            }
+
+            if ((show_view_title_overlay == title_overlay_t::NEVER) ||
+                ((show_view_title_overlay == title_overlay_t::MOUSE) &&
+                 (tmp_view != last_title_overlay)))
+            {
+                return;
+            }
+
+            ol->render(fb, damage);
+        });
+
+        wf::scale_transformer_t::padding pad;
+        const std::string& pos = title_position;
+        if (pos == "top")
+        {
+            ol->pos = view_title_overlay_t::position::TOP;
+            pad.top = ol->get_size().height;
+        } else if (pos == "bottom")
+        {
+            ol->pos    = view_title_overlay_t::position::BOTTOM;
+            pad.bottom = ol->get_size().height;
+        }
+
+        tr->add_overlay(cb, 1, pad);
+    }
+},
+
+mouse_update{[this] (auto)
+    {
+        update_title_overlay_mouse();
+    }
+}
+
+{}
+
+void scale_show_title_t::init(wf::output_t *output)
+{
+    this->output = output;
+    output->connect_signal("scale-filter", &view_filter);
+    output->connect_signal("scale-transformer-added", &add_title_overlay);
+    output->connect_signal("scale-end", &scale_end);
+}
+
+void scale_show_title_t::fini()
+{
+    output->disconnect_signal(&view_filter);
```

Comment:
no need for manually disconnecting actually, because the signal_connection_t objects will do this automatically.